### PR TITLE
[SPARK-39824][PYTHON][PS] Introduce index where and putmask func in pyspark

### PR DESCRIPTION
**Pull request by bzhaoopenstack**
### What changes were proposed in this pull request?

Add more pyspark pandas Index func which is similar with pandas.

### Why are the changes needed?

Add where and putmask which is very similar func into pyspark pandas

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

```
        >>> idx = ps.Index(['car', 'bike', 'train', 'tractor'])
        >>> idx
        Index(['car', 'bike', 'train', 'tractor'], dtype='object')
        >>> idx.where(idx.isin(['car', 'train']), 'other')
        Index(['car', 'other', 'train', 'other'], dtype='object')
```

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @zhengruifeng @xinrong-meng @itholic FYI

File: python/pyspark/pandas/tests/indexes/test_base.py [hunk 1]
```diff
@@ -2511,6 +2511,36 @@ def test_drop_level(self):
...
+    
+        # where and putmask with isin func
+        self.assert_eq(
+            pidx.where(pidx.isin([1, 2])),
+            psidx.where(psidx.isin([1, 2]))
+        )
+        self.assert_eq(
+            pidx.putmask(pidx.isin([1, 2]), 99),
+            psidx.putmask(psidx.isin([1, 2]), 99)
+        )
```
Comment (itholic) [inline python/pyspark/pandas/tests/indexes/test_base.py:2536]: Shall we have a bit more test for complex case ??

e.g.
```python
(pidx + 1).putmask((pidx + 1).isin([1, 2]), 99),
(psidx + 1).putmask((psidx + 1).isin([1, 2]), 99)
```

**Comment by itholic** [comment]
Oh btw, we have `[PS]` tag for the PR title.

Can we also add PS to the title ? (and your other open PRs such as https://github.com/apache/spark/pull/37044, https://github.com/apache/spark/pull/37232, https://github.com/apache/spark/pull/37234 as well :-))

**Comment by bzhaoopenstack** [comment]
> Oh btw, we can use `[PS]` tag for the PR title related to pandas-on-Spark changes.
> 
> Can we also add `[PS]` to the title ? (and your other open PRs such as #37044, #37232, #37234 as well :-))

OK. Sorry that not familiar with Spark PR contribution protocol

File: python/pyspark/pandas/tests/indexes/test_base.py (hunk 1 above)
Comment (bzhaoopenstack) [reply python/pyspark/pandas/tests/indexes/test_base.py:2536]: OK.  I will add into UTs.

```
>>> (pidx + 1).putmask((pidx+1).isin([1,2]), 99)
Int64Index([99, 3, 4, 5], dtype='int64')
>>> (psidx + 1).putmask((psidx + 1).isin([1, 2]), 99)
/home/spark/spark/python/pyspark/pandas/utils.py:976: PandasAPIOnSparkAdviceWarning: `to_list` loads all data into the driver's memory. It should only be used if the resulting list is expected to be small.
  warnings.warn(message, PandasAPIOnSparkAdviceWarning)
/home/spark/spark/python/pyspark/pandas/indexes/base.py:636: UserWarning: We recommend using `Int64Index.to_numpy()` instead.
  warnings.warn("We recommend using `{}.to_numpy()` instead.".format(type(self).__name__))
/home/spark/spark/python/pyspark/pandas/utils.py:976: PandasAPIOnSparkAdviceWarning: `to_numpy` loads all data into the driver's memory. It should only be used if the resulting NumPy ndarray is expected to be small.
  warnings.warn(message, PandasAPIOnSparkAdviceWarning)
/home/spark/spark/python/pyspark/pandas/indexes/base.py:636: UserWarning: We recommend using `Int64Index.to_numpy()` instead.
  warnings.warn("We recommend using `{}.to_numpy()` instead.".format(type(self).__name__))
/home/spark/spark/python/pyspark/pandas/utils.py:976: PandasAPIOnSparkAdviceWarning: `to_numpy` loads all data into the driver's memory. It should only be used if the resulting NumPy ndarray is expected to be small.
  warnings.warn(message, PandasAPIOnSparkAdviceWarning)
/home/spark/spark/python/pyspark/pandas/internal.py:1573: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  fields = [
/home/spark/spark/python/pyspark/sql/pandas/conversion.py:486: FutureWarning: iteritems is deprecated and will be removed in a future version. Use .items instead.
  for column, series in pdf.iteritems():
Int64Index([99, 3, 4, 5], dtype='int64')

```

File: python/pyspark/pandas/indexes/base.py [hunk 2]
```diff
@@ -2666,6 +2666,93 @@ def inferred_type(self) -> str:
...
         return lib.infer_dtype([self.to_series().head(1).item()])
 
+    def putmask(
+            self,
+            mask: "Index",
+            value: Any
+    ) -> "Index":
+        """
+        Return a new Index of the values set with the mask.
+
```
Comment (zhengruifeng) [inline python/pyspark/pandas/indexes/base.py:2676]:
```
        .. versionadded:: 3.4.0
```

File: python/pyspark/pandas/indexes/base.py [hunk 3]
```diff
@@ -2666,6 +2666,93 @@ def inferred_type(self) -> str:
...
+        return Index(new_values)
+    
+    def where(
+        self,
+        cond: "Index",
+        other: Any = np.nan
+    ) -> "Index":
+        """
+        Replace values where the condition is False.
+
```
Comment (zhengruifeng) [inline python/pyspark/pandas/indexes/base.py:2725]: ditto

**Comment by zhengruifeng** [comment]
new methods should also be listed in python/docs/source/reference/pyspark.pandas/indexing.rst

**Comment by zhengruifeng** [comment]
this failure needs be fixed
```
======================================================================
ERROR [0.436s]: test_missing (pyspark.pandas.tests.indexes.test_base.IndexesTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/spark/spark/python/pyspark/pandas/tests/indexes/test_base.py", line 508, in test_missing
    getattr(psdf.set_index(["a", "b"]).index, name)()
TypeError: putmask() missing 2 required positional arguments: 'mask' and 'value'
----------------------------------------------------------------------
```

as to the linter failure, you may just run `dev/reformat-python`

**Comment by zhengruifeng** [comment]
also cc @Yikun

File: python/pyspark/pandas/indexes/base.py (hunk 2 above)
Comment (xinrong-meng) [reply python/pyspark/pandas/indexes/base.py:2676]: Also
```
    Parameters
    ----------
```
please

File: python/pyspark/pandas/indexes/base.py [hunk 4]
```diff
@@ -2666,6 +2666,93 @@ def inferred_type(self) -> str:
...
+        -------
+        Index
+    
+        See Also
+        --------
+        pandas.index.putmask : Changes elements of an array
+            based on conditional and input values.
+        """
+        # validate mask
+        mask = np.asarray(mask.tolist(), dtype=bool)  # type: ignore[assignment]
```
Comment (xinrong-meng) [inline python/pyspark/pandas/indexes/base.py:2687]: Is it safe to assume `mask` fits into the memory?

File: python/pyspark/pandas/indexes/base.py (hunk 3 above)
Comment (bzhaoopenstack) [reply python/pyspark/pandas/indexes/base.py:2725]: OK

**Comment by bzhaoopenstack** [comment]
> this failure needs be fixed
> 
> ```
> ======================================================================
> ERROR [0.436s]: test_missing (pyspark.pandas.tests.indexes.test_base.IndexesTest)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/__w/spark/spark/python/pyspark/pandas/tests/indexes/test_base.py", line 508, in test_missing
>     getattr(psdf.set_index(["a", "b"]).index, name)()
> TypeError: putmask() missing 2 required positional arguments: 'mask' and 'value'
> ----------------------------------------------------------------------
> ```
> 
> as to the linter failure, you may just run `dev/reformat-python`

Hi Zheng, Thank you very much. But I think I need some helps about the MultiIndex putmask.
I didn't find any usecase about MultiIndex putmask, and the behaviors between pandas and numpy are different too. So I'm still not verify the putmask func I introduced work well. Could you pls show some usecases for helping this?

BTW, This due to I didn't remove the 'putmask' in MultiIndex missing List. ;-).

File: python/pyspark/pandas/indexes/base.py (hunk 4 above)
Comment (bzhaoopenstack) [reply python/pyspark/pandas/indexes/base.py:2687]: I think that's OK just for a Index mask. Or could u pls tell me more about you feel it not safe? We can discuss further for making sure we are on the same page. ;-)

**Comment by zhengruifeng** [comment]
@bzhaoopenstack 

> I didn't find any usecase about MultiIndex putmask

does pandas support `MultiIndex putmask`

>  the behaviors between pandas and numpy are different too.

do you mean the `MultiIndex`?

**Comment by bzhaoopenstack** [comment]
> @bzhaoopenstack
> 
> > I didn't find any usecase about MultiIndex putmask
> 
> does pandas support `MultiIndex putmask`

>>> From the release doc and API doc, yeah, it claims that MultiIndex supports putmask(doesn't support where), but no any sample. So it's hard to verify the functionality for me, as I never call multiIndex.putmask with pandas successfully.

> 
> > the behaviors between pandas and numpy are different too.
> 
> do you mean the `MultiIndex`?
>>> Yeah, from my exploration, the numpy.putmask behavior is different with pandas MultiIndex.putmask. As those can success in numpy, but fail in pandas.

**Comment by zhengruifeng** [comment]
> Yeah, from my exploration, the numpy.putmask behavior is different with pandas MultiIndex.putmask. As those can success in numpy, but fail in pandas.

I guess we don't care the difference between pandas and numpy. We'd better make the behavior of PS(Pandas API on Spark) the same as Pandas.

**Pull request closed without merging**